### PR TITLE
Versions: keep type of version in sync with the project

### PR DESCRIPTION
**Pull request by stsewd**
So, there are a couple of problems here:

- Versions have been created with the type set as unknown, and where never updated to the correct type. I was able to track down this to projects created before 2018, so probably an old bug allowed this to happen.
- Stable machine versions are basically an alias for the latest stable version of the project, and they can be a branch or a tag, this means that if this version is a branch, so will be the "stable" version, same for tags. Currently, we aren't updating the version type, if the stable version was created as a branch, it will remain as a branch, even if the latest stable version is a tag, and vice versa.
- Latest machine versions are basically an alias for the default branch of the project, they are always branches. Since we allow having non-machine latest versions, they can be a branch or a tag. If this version was changed back to be a non-machine version, currently it will remain as branch or tag, this was incorrect.

Our new clone/checkout pattern relies on the type of the version always being correct, so it was failing for some projects.

- Fixes https://github.com/readthedocs/readthedocs.org/issues/10600
- Fixes https://github.com/readthedocs/readthedocs.org/issues/10601

After deploy, we need to clean up the invalid versions (versions with "unknown" type). We can do that by:

- Updating all latest machine versions to be branches.

  ```python
  python Version.objects.filter(type="unknown", machine=True).update(type="branch")
  ```

- Re-evaluate the stable version of all projects that have a machine stable version, so they can have the correct type.

  ```python
  for project in Project.objects.filter(versions__machine=True).distinct():
      project.update_stable_version()
  ```

- Check all remaining active versions with "unknown" type, and update them to be branches or tags (324). We can use a simple regex to see the version identifier looks like a commit.

  ```python
  python vesions = Version.objects.filter(type="unknown", active=True)
  commit_regex = re.compile(r"^([a-f0-9]{40})|([a-f0-9]{7})$")
  for version in versions:
    if commit_regex.match(version.identifier):
      version.type = "tag"
    else:
      version.type = "branch"
      version.save()
   ```

### CODE REVIEW

**Review by humitos — COMMENTED** [review]
This is great! Thanks for fixing this issue. I'm happy to deploy this next week and continue moving forward with our new Git+Clone+Fetch implementation 🚀

File: readthedocs/rtd_tests/tests/test_sync_versions.py [hunk 1]
```diff
@@ -219,6 +219,94 @@ def test_delete_version(self):
...
+            tags_data=[],
+        )
+
+        self.pip.update_stable_version()
+        stable_version = self.pip.get_stable_version()
+        self.assertEqual(stable_version.type, BRANCH)
+        self.assertEqual(stable_version.identifier, "2.0")
+        self.assertEqual(stable_version.verbose_name, "stable")
+
+        original_stable = self.pip.get_original_stable_version()
```
Comment (humitos) [inline readthedocs/rtd_tests/tests/test_sync_versions.py:261]: I doubted what was the meaning of "original stable", so I went to the docstring: https://github.com/readthedocs/readthedocs.org/blob/bbb200ba2e42d032843e3f8da170c61b76af514e/readthedocs/projects/models.py#L1167-L1172

but it's not super useful. Do you know what this mean? What is the difference between the "original stable" (`get_original_stable_version`) and the "stable" (`get_stable_version`)?

**Review by humitos — APPROVED** [review]
This looks good!

I left a question about "stable" and "original stable" that I'd like to understand more so we can update the docstring of that method and make it clearer.

File: readthedocs/rtd_tests/tests/test_sync_versions.py (hunk 1 above)
Comment (stsewd) [reply readthedocs/rtd_tests/tests/test_sync_versions.py:261]: When stable is machine created, it's basically an alias for the latest stable version (like 2.2), sometimes we need to point to that version, that's the "original" one.

Comment (humitos) [reply readthedocs/rtd_tests/tests/test_sync_versions.py:261]: oh, I see. Would you mind adding that comment in the docstring of that method? 🙏🏼

**Pull request merged**